### PR TITLE
tailcfg: make node.Hostinfo a HostinfoView

### PR DESCRIPTION
**Pull request by maisem**
Signed-off-by: Maisem Ali <maisem@tailscale.com>

### CODE REVIEW

File: tailcfg/tailcfg.go [hunk 1]
```diff
@@ -454,6 +459,126 @@ type Hostinfo struct {
 	//       require changes to Hostinfo.Equal.
 }
 
+func (hi *Hostinfo) View() HostinfoView { return HostinfoView{hi} }
+
+type HostinfoView struct {
```
Comment (bradfitz) [inline tailcfg/tailcfg.go:464]: needs docs. This is the first time we're using Views in the oss tree, so people will be confused by them and ж

File: types/views/views.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+import (
+	"inet.af/netaddr"
+	"tailscale.com/net/tsaddr"
+)
+
+// StringsSlice is a read-only accessor for a slice of strings.
+type StringSlice struct {
+	// ж represents the underlying value.
+	// It is considered a lower-case letter so it is not exported.
```
Comment (bradfitz) [inline types/views/views.go:17]: Not worth saying. If anybody tries it they'll learn that soon enough.

I'd say instead:

```go
     // It is named distinctively to make you think of how dangerous it is to escape
     // to callers. You must not let callers be able to mutate it.
```

File: types/views/views.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+// Package views provides read-only accessors for commonly used
+// value types.
+package views
+
+import (
+	"inet.af/netaddr"
+	"tailscale.com/net/tsaddr"
+)
+
+// StringsSlice is a read-only accessor for a slice of strings.
```
Comment (bradfitz) [inline types/views/views.go:14]: String singular

File: ipn/ipnlocal/local.go [hunk 4]
```diff
@@ -2972,7 +2973,9 @@ func peerAPIBase(nm *netmap.NetworkMap, peer *tailcfg.Node) string {
 		}
 	}
 	var p4, p6 uint16
-	for _, s := range peer.Hostinfo.Services {
+	svcs := peer.Hostinfo.Services()
+	for i, l := 0, svcs.Len(); i < l; i++ {
```
Comment (bradfitz) [inline ipn/ipnlocal/local.go:2977]: i, n

File: ipn/ipnlocal/local.go [hunk 5]
```diff
@@ -3178,7 +3181,9 @@ func exitNodeCanProxyDNS(nm *netmap.NetworkMap, exitNodeID tailcfg.StableNodeID)
 		if p.StableID != exitNodeID {
 			continue
 		}
-		for _, s := range p.Hostinfo.Services {
+		services := p.Hostinfo.Services()
+		for i, l := 0, services.Len(); i < l; i++ {
```
Comment (bradfitz) [inline ipn/ipnlocal/local.go:3185]: etc

File: tailcfg/tailcfg.go [hunk 6]
```diff
@@ -454,6 +459,120 @@ type Hostinfo struct {
...
+	}
+	hi := &Hostinfo{}
+	if err := json.Unmarshal(b, hi); err != nil {
+		return err
+	}
+	v.ж = hi
+	return nil
+}
+
+// Valid reports whether the underlying struct is not nil.
```
Comment (bradfitz) [inline tailcfg/tailcfg.go:492]: technically, structs are never nil.  "underlying value"?

File: tailcfg/tailcfg.go [hunk 7]
```diff
@@ -454,6 +459,120 @@ type Hostinfo struct {
...
+func (v HostinfoView) Valid() bool { return v.ж != nil }
+
+// AsStruct returns a deep-copy of the underlying struct.
```
Comment (bradfitz) [inline tailcfg/tailcfg.go:495]: value?

File: tailcfg/tailcfg.go [hunk 8]
```diff
@@ -454,6 +459,120 @@ type Hostinfo struct {
...
+func (v HostinfoView) AsStruct() *Hostinfo { return v.ж.Clone() }
+
+// Clone creates a shallow-copy of the HostinfoView.
+// This is safe as the underlying struct is considered immutable.
+func (v *HostinfoView) Clone() *HostinfoView { return &HostinfoView{v.ж} }
```
Comment (bradfitz) [inline tailcfg/tailcfg.go:500]: wait, why? I'm lost.

And why ever a pointer to a view?

File: tailcfg/tailcfg.go [hunk 9]
```diff
@@ -454,6 +459,120 @@ type Hostinfo struct {
...
+func (v HostinfoView) Services() ServiceSlice {
+	return ServiceSliceOf(v.ж.Services)
+}
+
+func (v HostinfoView) NetInfo() NetInfoView { return v.ж.NetInfo.View() }
+
+// ServiceSlice is a read-only accessor for a slice of Services
+type ServiceSlice struct {
+	// ж represents the underlying value.
+	// It is considered a lower-case letter so it is not exported.
```
Comment (bradfitz) [inline tailcfg/tailcfg.go:532]: same

File: tailcfg/tailcfg.go (hunk 8 above)
Comment (maisem) [reply tailcfg/tailcfg.go:500]: Oh this was me fighting with cloner. How do I make it play nice with views?

File: tailcfg/tailcfg.go (hunk 9 above)
Comment (maisem) [reply tailcfg/tailcfg.go]: Done

File: tailcfg/tailcfg.go (hunk 8 above)
Comment (maisem) [reply tailcfg/tailcfg.go]: Fixed cloner

File: tailcfg/tailcfg.go (hunk 7 above)
Comment (maisem) [reply tailcfg/tailcfg.go]: Done

File: tailcfg/tailcfg.go (hunk 6 above)
Comment (maisem) [reply tailcfg/tailcfg.go]: Done.

File: ipn/ipnlocal/local.go (hunk 5 above)
Comment (maisem) [reply ipn/ipnlocal/local.go]: Done.

File: ipn/ipnlocal/local.go (hunk 4 above)
Comment (maisem) [reply ipn/ipnlocal/local.go]: Done.

File: types/views/views.go (hunk 3 above)
Comment (maisem) [reply types/views/views.go]: Done.

File: types/views/views.go (hunk 2 above)
Comment (maisem) [reply types/views/views.go]: Done.

File: tailcfg/tailcfg.go (hunk 1 above)
Comment (maisem) [reply tailcfg/tailcfg.go:467]: Done

File: tailcfg/tailcfg.go [hunk 10]
```diff
@@ -454,6 +459,126 @@ type Hostinfo struct {
...
+// AsSlice returns a copy of underlying slice.
+func (v ServiceSliceView) AsSlice() []Service {
+	return v.Append(v.ж[:0:0])
+}
+
+type StringSliceView struct {
+	ж []string
+}
+
+func NewStringSliceView(x []string) StringSliceView { return StringSliceView{x} }
```
[the remark this answers is not in the document]
Comment (maisem) [reply tailcfg/tailcfg.go]: Done.

File: tailcfg/tailcfg.go [hunk 11]
```diff
@@ -532,6 +659,9 @@ func (ni *NetInfo) portMapSummary() string {
 	return prefix + conciseOptBool(ni.UPnP, "U") + conciseOptBool(ni.PMP, "M") + conciseOptBool(ni.PCP, "C")
 }
 
+// View returns a read-only accessor for the NetInfo object.
```
Comment (bradfitz) [inline tailcfg/tailcfg.go:662]: Go doesn't really ever use the word "object". Maybe "a read-only accessor for ni." ?

File: tailcfg/tailcfg.go [hunk 12]
```diff
@@ -454,6 +459,128 @@ type Hostinfo struct {
...
+func (v HostinfoView) BackendLogID() string       { return v.ж.BackendLogID }
+func (v HostinfoView) OS() string                 { return v.ж.OS }
+func (v HostinfoView) OSVersion() string          { return v.ж.OSVersion }
+func (v HostinfoView) Package() string            { return v.ж.Package }
+func (v HostinfoView) DeviceModel() string        { return v.ж.DeviceModel }
+func (v HostinfoView) Hostname() string           { return v.ж.Hostname }
+func (v HostinfoView) ShieldsUp() bool            { return v.ж.ShieldsUp }
+func (v HostinfoView) ShareeNode() bool           { return v.ж.ShareeNode }
+func (v HostinfoView) GoArch() string             { return v.ж.GoArch }
+func (v HostinfoView) Equal(h2 HostinfoView) bool { return v.ж.Equal(h2.ж) }
```
Comment (bradfitz) [inline tailcfg/tailcfg.go:509]: v2? there's no h or h1.

File: tailcfg/tailcfg.go [hunk 13]
```diff
@@ -454,6 +459,128 @@ type Hostinfo struct {
 	//       require changes to Hostinfo.Equal.
 }
 
+// View returns a read-only accessor for the Hostinfo object.
```
Comment (bradfitz) [inline tailcfg/tailcfg.go:462]: for hi

**Review by bradfitz — APPROVED** [review]

**Pull request merged**